### PR TITLE
Update packaging docs.

### PR DESCRIPTION
**Pull request by associahedron**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/daml/reference/packages.rst [hunk 1]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
 DAML archives
 *************
 
-When a DAML project is build with ``daml build``, build artifacts are generated in the hidden
-directory ``.daml/dist/`` relative to the project root directory. The main build artifact of a
-project is the `DAML archive`, recognized by the ``.dar`` file ending. DAML archives are platform
-independent. They can be deployed on a ledger (see :ref:`deploy <deploy-ref_overview>`) or can be
-imported into other projects as a package dependency.
+When a DAML project is compiled, the compiler produces a `DAML archive`. These are platform-independent packages of compiled DAML code, that can be uploaded to a DAML ledger or imported in other DAML projects.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]:
```suggestion
When a DAML project is compiled, the compiler produces a `DAML archive`. These are platform-independent packages of compiled DAML code that can be uploaded to a DAML ledger or imported in other DAML projects.
```

File: docs/source/daml/reference/packages.rst [hunk 2]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+.. code-block:: daml
 
-.. code-block:: yaml
+  import Foo
+
+Sometimes you will have multiple packages with the same exported module name. In that case, you must also specify which package the module comes from, as follows:
+
+.. code-block:: daml
+
+  import "foo" Foo
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: We should document how you can rename modules if you have the same name but different versions so you cannot use package-qualified imports to disambiguate. Totally fine to do that in a separate PR. This is also not specific to dependencies or data-dependencies so I wonder if it would make sense to put this into a separate “Handling colliding module names” section at the end or somethin glike that.

File: docs/source/daml/reference/packages.rst [hunk 3]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+By default, all modules of ``foo`` are made available when importing ``foo`` as a dependency. To control which modules of ``foo`` get exported, you may add an ``exposed-modules`` field in the ``daml.yaml`` file for ``foo``:
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]:
```suggestion
By default, all modules of ``foo`` are made available when importing ``foo`` as a dependency. To limit which modules of ``foo`` get exported, you may add an ``exposed-modules`` field in the ``daml.yaml`` file for ``foo``:
```
Makes it a bit clearer that you only need this if you don’t want to export everything.

File: docs/source/daml/reference/packages.rst [hunk 4]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
-  sdk-version: 0.0.0
-  name: foo
-  source: daml
-  version: 1.0.0
   exposed-modules:
-    - Some.Module
-    - Some.Other.Module
+  - Foo
+
+**Important Limitation:** For DAML archive imports via ``dependencies`` to work, the archive should be compiled with the same DAML SDK version. Otherwise, compilation will fail. If matching the SDK version is not possible, see the next section.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I expect that at least for now where libraries don’t really exist, most users will actually want `data-dependencies` not `dependencies` so the warning here at the end might be a bit late. What do you think of adding a small section at the beginning that compares them and maybe swapping the order of `dependencies` and `data-dependencies`?

Happy to leave that out of this PR.

File: docs/source/daml/reference/packages.rst [hunk 5]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+Importing a DAML archive via data-dependencies
+**********************************************
+
+A secondary method for importing a DAML archive, which can be used when the DAML SDK versions do not match, is to import a DAML archive via the ``data-dependencies`` field in ``daml.yaml``:
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: The way this is written it sounds like maybe I should start out using `dependencies` and then switch to `data-dependencies` once I upgrade SDK versions. I think this is a bit misleading. At least in my view, you should always use data-dependencies once you refer to something where the package id is fixed (which includes everything that has already been uploaded to the ledger). The only case where you should use `dependencies` is when you don’t care about the package id and are free to recompile it at some point, potentially in a different SDK version.

File: docs/source/daml/reference/packages.rst [hunk 6]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
-    - /home/johndoe/bar/.daml/dist/bar.dar
+  - daml-prim
+  - daml-stdlib
   data-dependencies:
-    - /home/jondoe/bar-0.0.0/.daml/dist/bar-0.0.0.dar
+  - ../foo/foo.dar
 
-Modules from data dependencies can be imported as usual, but need to be qualified by the
-(generated) instances package:
+You can also import a ``.dalf`` file via data-dependencies.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I would leave this out. Shipping DALFs around on their own is not something I want anyone to do and I don’t think there are really good reasons for it. as soon as your DALF is not self-contained you also manually have to ensure that all your dependencies are there which becomes very messy.

File: docs/source/daml/reference/packages.rst [hunk 7]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
 
-.. code-block:: daml
+When importing packages this way, the DAML compiler will attempt to reconstruct the DAML interface from the compiled DAML-LF binaries included in the DAML archive.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: This sounds a bit too much like it’s describing the implementation rather than what I care about as a user. Not quite sure what a better phrasing is.

File: docs/source/daml/reference/packages.rst [hunk 8]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+Compared to dependencies, there are certain disadvantages to this method.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]:
```suggestion
To allow `data-dependencies` to work across SDK versions, they have to abstract over some details which are not compatible across SDK versions. This means that there are some DAML features that cannot be recovered when using `data-dependencies`.
```
Sounds a bit less negative, happy to change the wording. This is just a suggestion :slightly_smiling_face:

File: docs/source/daml/reference/packages.rst [hunk 9]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+The first disadvantage is that the reconstruction process used for data-dependencies is slower than the direct import process used for dependencies, so it will negatively affect the speed of compilation.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I would leave this out. This is mainly an artifact of #4413. Once that is implemented, you only pay for this once  and it’s not really a big deal.

File: docs/source/daml/reference/packages.rst [hunk 10]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+The second disadvantage, which has far-reaching consequences, is that not everything can be perfectly reconstructed via data-dependencies. In particular:
+
+#. Export lists cannot be reconstructed, so imports via data-dependencies can access definitions that were originally hidden. This means it is up to the importing module to respect the data encapsulation of the original module. On the positive side, the encapsulation can also be ignored on purpose, to facilitate upgrades of DAML models to newer SDK versions.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I think it’s worth pointing out somewhere here that this is not any worse than DAML-LF and therefore the ledger API which also does not support data abstraction (:crying_cat_face:). Currently it sounds a bit too scary imho.

File: docs/source/daml/reference/packages.rst [hunk 11]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+#. Certain advanced type system features also cannot be reconstructed, as they are erased in the process of compiling DAML LF binaries. This includes the ``DataKinds``, ``DeriveGeneric``, and ``FunctionalDependencies`` extensions from GHC. This may result in some definitions being unavailable when importing a module that uses these advanced features.
+
+#. Prior to DAML LF version 1.8, typeclasses could not be reconstructed from DAML archives. This means if you have an archive that is compiled with an older version of DAML LF, typeclasses and typeclass instances will not be carried over via data-dependencies, and you will not be able to call functions that rely on typeclass instances.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I think it’s worth mentioning here that this includes all template operations since most users probably have no clue that those involve typeclasses.

File: docs/source/daml/reference/packages.rst [hunk 12]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+Compared to dependencies, there are certain disadvantages to this method.
+
+The first disadvantage is that the reconstruction process used for data-dependencies is slower than the direct import process used for dependencies, so it will negatively affect the speed of compilation.
+
+The second disadvantage, which has far-reaching consequences, is that not everything can be perfectly reconstructed via data-dependencies. In particular:
+
+#. Export lists cannot be reconstructed, so imports via data-dependencies can access definitions that were originally hidden. This means it is up to the importing module to respect the data encapsulation of the original module. On the positive side, the encapsulation can also be ignored on purpose, to facilitate upgrades of DAML models to newer SDK versions.
+
+#. Certain advanced type system features also cannot be reconstructed, as they are erased in the process of compiling DAML LF binaries. This includes the ``DataKinds``, ``DeriveGeneric``, and ``FunctionalDependencies`` extensions from GHC. This may result in some definitions being unavailable when importing a module that uses these advanced features.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I think `DataKinds` don’t work in general anyway? We special case a very small number of kinds, primarily generics, in https://github.com/digital-asset/daml/blob/master/compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs#L1606 but I’m not sure it works at all outside of that.

For FunctionalDepencies, I believe things work as long as we can reexport the typeclass?

File: docs/source/daml/reference/packages.rst (hunk 11 above)
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I would switch this section and the one before. I expect basically noone to use the advanced typelevel stuff but typeclasses are more important and at least for a while we’ll have people trying to use DAML-LF < 1.8.

File: docs/source/daml/reference/packages.rst [hunk 13]
```diff
@@ -12,75 +12,94 @@ This page gives reference information on DAML package dependencies:
...
+
+#. When possible, typeclass instances will be reconstructed using the typeclass definitions from dependencies (such as the typeclass definitions from ``daml-stdlib``). But if the typeclass methods or signature has changed, you will get an instance for a reconstructed typeclass instead, which will not interoperate with code from dependencies. So this is something to keep in mind when typeclass definitions have changed.
 
-  import "instances-bar" Foo
+.. TODO (#4932): Add warnings for advanced features that aren't supported, and add a comment on bullet #2.
 
+Given this long list of disadvantages, data-dependencies are a tool that is only recommended when dependencies cannot be used. In particular, data-dependencies should only be used to interface with deployed code on a ledger, such as to interact with a deployed DAML model or to upgrade of a deployed DAML model. See the :ref:`upgrade documentation <upgrade-overview>` for more details on the latter.
```
Comment (cocreature) [inline docs/source/daml/reference/packages.rst]: I agree with this but I think it’s worded a bit too strongly. As mentioned above, I expect that right now, data-dependencies will see far more usage than dependencies so I would go the other way around “use data-dependencies unless you have a library that you can recompile”.

File: docs/source/daml/reference/packages.rst (hunk 1 above)
Comment (associahedron) [reply docs/source/daml/reference/packages.rst]: :'( English is so ambiguous.

File: docs/source/daml/reference/packages.rst (hunk 2 above)
Comment (associahedron) [reply docs/source/daml/reference/packages.rst]: is this just `import "foo" Foo as Foo1`, or something else? I agreed that dealing with conflicting module names should move to a separate section.

Comment (cocreature) [reply docs/source/daml/reference/packages.rst]: No you need to specify something like `--package foo-1.0.0 (Foo as Foo1)` in your `daml.yaml` in that case.

Comment (cocreature) [reply docs/source/daml/reference/packages.rst]: See https://github.com/digital-asset/daml/blob/master/compiler/damlc/tests/src/DA/Test/Packaging.hs#L102

Comment (associahedron) [reply docs/source/daml/reference/packages.rst]: Oh, I see, this is for the case where you have two packages with the same name but different versions. This is terrible UX. Could we not change unit IDs to make package qualified imports work in this case?

Comment (associahedron) [reply docs/source/daml/reference/packages.rst]: (I mean, change the way unit IDs are chosen from LF packages, so they include the package version.)

Comment (cocreature) [reply docs/source/daml/reference/packages.rst]: GHC has fairly strictly rules about what can be a package name sadly. `foobar-1.0.0` is a valid unit id but it’s not a valid package name (the latter is what you use in package-qualified imports). I believe `foobar-1_0_0` would be but that doesn’t seem very satisfying. I looked into patching GHC at some point to allow the unit id in package-qualified imports and it looked fairly doable but then decided that I’m too lazy :slightly_smiling_face: I think for now, documenting the current state first and then seeing if we can improve upon it seems like the best option.

Comment (associahedron) [reply docs/source/daml/reference/packages.rst]: Sounds good 👍 

Maybe in the future the user could specify an alias for packages in `dependencies` or `data-dependencies`, e.g.

``` yaml
dependencies:
- daml-prim
- daml-stdlib
- package: foo-1.4.2.dar
  alias: foo-v1  
```

or whatever syntax is most appropriate

Comment (cocreature) [reply docs/source/daml/reference/packages.rst]: ah yeah, I thought about this in the past but then forgot about it. That sounds like a reasonably simple option :+1:

Comment (associahedron) [reply docs/source/daml/reference/packages.rst]: I added an issue so we don't forget about it:
https://github.com/digital-asset/daml/issues/4948
